### PR TITLE
[anaconda]-auto install vuln pkgs from conda / pip

### PR DESCRIPTION
**Pull request by gauravsaini04**
**Dev container name**:
 
 * Anaconda
 
 **Description**:

This PR adds the following functionality : 

* Auto install vulnerable packages from conda channel / pip pkg managers

### CODE REVIEW

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
+
+# Find and install packages, if required, using conda channel or using pip package manager
+vulnerable_packages=( "pydantic=2.5.3" "joblib=1.3.1" "mistune=3.0.1" "werkzeug=3.0.3" "transformers=4.36.0" "pillow=10.3.0" "aiohttp=3.9.4" \
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:4]: This logic seems way too complex, discussing alternatives offline.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh (hunk 1 above)
Comment (gauravsaini04) [reply src/anaconda/.devcontainer/install_vulnerable_packages.sh:4]: have corrected based on comments.

**Review by Stam098c — COMMENTED** [review]
🙏💓☺️👍😊

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+# Function to compare semver versions
+compare_semver() {
+    # Split versions into arrays
+    IFS='.' read -r -a version1 <<< "$1"
+    IFS='.' read -r -a version2 <<< "$2"
+
+    comparison=""
+    # Compare MAJOR version
+    if (( ${version1[0]} > ${version2[0]} )); then
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:32]: Why do we need this comparison? 

In the reference sample 👇 , it takes care of comparing the semvers. It finds the max out of required and current in `GREATER_VERSION` and only installs the patch if installed version is not greater or equal to the required version.

```
CURRENT_VERSION=$(git --version | awk '{print $3}')

    REQUIRED_GIT_VERSION="2.45.1"
    GREATER_VERSION=$((echo ${CURRENT_VERSION}; echo ${REQUIRED_GIT_VERSION}) | sort -V | tail -1)

    # Check if the git verson installed is greater or equal to the required git version
    if [ "${CURRENT_VERSION}" != "${GREATER_VERSION}" ]; then
        echo "Git version ${CURRENT_VERSION} installed by the Ubuntu base image is not greater or equal to ${REQUIRED_GIT_VERSION}"
        echo "Installing git from source for ${REQUIRED_GIT_VERSION}..."
        .....
    fi
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    IFS='=' read -ra parts <<< "${vulnerable_packages[$i]}"
+    # Assign the parts to the 2D array
+    packages_array[$i,0]=${parts[0]}
+    packages_array[$i,1]=${parts[1]}
+done
+
+for ((i=0; i<rows; i++)); do
+    CURRENT_VERSION=$(pip show "${packages_array[$i,0]}" | grep '^Version:' | awk '{print $2}')
+    if [[ -z "$CURRENT_VERSION" ]]; then
+        echo "No version for ${packages_array[$i,0]} found in upstream i.e. base image."
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:27]: As we are patching the existing upstream image, ideally this should never happen.

In future, if our tests start failing then it's an indication for us to clean up this patch. Hence, can we remove it?

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            uniq | \
+            tail -n 2 | \
+            head -n 1)
+        if [[ -z "$CONDA_VERSION" ]]; then
+            echo "No version for ${packages_array[$i,0]} found in conda channel."
+            CONDA_VERSION="0"
+        fi
+        GREATER_VERSION_B=$((echo ${REQUIRED_VERSION}; echo ${CONDA_VERSION}) | sort -V | tail -1)
+        if [[ $CONDA_VERSION == $GREATER_VERSION_B ]]; then
+            echo -e "Conda Version: ${CONDA_VERSION} is greater than or equal to the required version: ${REQUIRED_VERSION}. \n";
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:50]:
```suggestion
            echo -e "Found version ${CONDA_VERSION} in the conda channel which is greater than or equal to the required version: ${REQUIRED_VERSION}. \n";
```

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            echo "Installing ${packages_array[$i,0]} from source from conda channel for ${REQUIRED_VERSION}..."
+            conda install "${packages_array[$i,0]}==${CONDA_VERSION}"
+        elif [[ $REQUIRED_VERSION == $GREATER_VERSION_B ]]; then 
+            echo -e "Required version: ${REQUIRED_VERSION} is greater than the conda version: ${CONDA_VERSION}. \n";
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:54]:
```suggestion
            echo -e "Required version ${REQUIRED_VERSION} is greater than the version found in the conda channel  ${CONDA_VERSION}. \n";
```

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            echo "Installing ${packages_array[$i,0]} from source from pip package manager for ${REQUIRED_VERSION}..."
+            python3 -m pip install --upgrade "${packages_array[$i,0]}==${REQUIRED_VERSION}"
+        fi
+    fi
```
Comment (samruddhikhandale) [inline src/anaconda/.devcontainer/install_vulnerable_packages.sh:58]: The script you provided is already quite optimized. However, there are a few minor improvements that can be made:

1. Use `pip show --no-index --disable-pip-version-check` to speed up the pip command.
2. Use `conda search --override-channels -c "$channel_name"` to speed up the conda command.
3. Use `python3 -m pip install --upgrade --no-cache-dir` to speed up the pip install command.

Can we make these changes?

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh (hunk 6 above)
Comment (gauravsaini04) [reply src/anaconda/.devcontainer/install_vulnerable_packages.sh:54]: done

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh (hunk 5 above)
Comment (gauravsaini04) [reply src/anaconda/.devcontainer/install_vulnerable_packages.sh]: done

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh (hunk 4 above)
Comment (gauravsaini04) [reply src/anaconda/.devcontainer/install_vulnerable_packages.sh]: done

File: src/anaconda/.devcontainer/install_vulnerable_packages.sh (hunk 3 above)
Comment (gauravsaini04) [reply src/anaconda/.devcontainer/install_vulnerable_packages.sh]: done

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

**Comment by Stam098c** [comment]
@gauravsaini04 @samruddhikhandale @Stam098c @eljog

**Pull request merged**